File: 693_Binary_Number_with_Alternating_Bits.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/binary-number-with-alternating-bits/
# Given a positive integer, check whether it has alternating bits: namely, if two adjacent bits will always
# have different values.

# Repeatedly divide by 2 where remainder is bit value.
# Time - O(log n)
# Space - O(1)

class Solution(object):
    def hasAlternatingBits(self, n):
        """
        :type n: int
        :rtype: bool
        """
        n, bit = divmod(n, 2)       # get first bit, avoids special case in while loop

        while n:

            if n % 2 == bit:        # next bit is same as current bit
                return False

            n, bit = divmod(n, 2)

        return True

File: java/693_Binary_Number_with_Alternating_Bits.java
/*
https://leetcode.com/problems/binary-number-with-alternating-bits/
Given a positive integer, check whether it has alternating bits: namely, if two adjacent bits will always have
different values.

Staring with least significant, check each digit is different from previous.
Time - O(log n)
Space - O(1)
*/

class Solution {
    public boolean hasAlternatingBits(int n) {

        int digit;
        int prev = -1;

        while (n > 0) {
            digit = n % 2;
            if (digit == prev)
                return false;
            n /= 2;
            prev = digit;
        }
        return true;
    }
}
